File: 1584_min_cost_to_connect_all_points/solution.rb
# LeetCode 1584 - Min Cost to Connect All Points
# https://leetcode.com/problems/min-cost-to-connect-all-points/

# @param {Integer[][]} points
# @return {Integer}
def min_cost_connect_points(points)
  n = points.length
  used = Array.new(n, false)
  dist = Array.new(n, 10**9)
  dist[0] = 0
  answer = 0
  n.times do
    u = (0...n).reject { |i| used[i] }.min_by { |i| dist[i] }
    used[u] = true
    answer += dist[u]
    (0...n).each do |v|
      next if used[v]
      d = (points[u][0] - points[v][0]).abs + (points[u][1] - points[v][1]).abs
      dist[v] = d if d < dist[v]
    end
  end
  answer
end

File: 1584_min_cost_to_connect_all_points/solution.php
<?php

class Solution {
    /**
     * @param Integer[][] $points
     * @return Integer
     */
    function minCostConnectPoints($points) {
        $n = count($points);
        $used = array_fill(0, $n, false);
        $dist = array_fill(0, $n, 1000000000);
        $dist[0] = 0;
        $answer = 0;

        for ($k = 0; $k < $n; $k++) {
            $u = -1;
            for ($i = 0; $i < $n; $i++) {
                if (!$used[$i] && ($u === -1 || $dist[$i] < $dist[$u])) {
                    $u = $i;
                }
            }
            $used[$u] = true;
            $answer += $dist[$u];
            for ($v = 0; $v < $n; $v++) {
                if (!$used[$v]) {
                    $d = abs($points[$u][0] - $points[$v][0]) + abs($points[$u][1] - $points[$v][1]);
                    if ($d < $dist[$v]) {
                        $dist[$v] = $d;
                    }
                }
            }
        }
        return $answer;
    }
}
